### PR TITLE
Both the DNA and the DAG RNN are completely deterministic even with dropout

### LINKED ISSUE #26: DAGRNN is not deterministic
Comparing two runs with the same random seed, I get the following output. The losses and scores should be identical. This means results are not reproducible.

```
dagrnn_regression regression f1f34085-82eb-4736-a1c7-f54662101143
epoch 0
100%|████████████████████████████████████████████████████████████████████████████████████████████████████| 119/119 [00:00<00:00, 175.69it/s]
100%|████████████████████████████████████████████████████████████████████████████████████████████████████| 119/119 [00:00<00:00, 492.48it/s]
train loss: 0.5581684708595276
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████| 38/38 [00:00<00:00, 485.24it/s]
validation loss: 2.508641004562378
100%|████████████████████████████████████████████████████████████████████████████████████████████████████| 413/413 [00:00<00:00, 595.18it/s]
100%|████████████████████████████████████████████████████████████████████████████████████████████████████| 128/128 [00:00<00:00, 606.40it/s]
train scores: {'RMSE': 0.5581684301395724}
test scores: {'RMSE': 2.508640777022769}
```

```
dagrnn_regression regression 4212bf85-6f9b-41ff-9855-37fde8a861a4
epoch 0
100%|████████████████████████████████████████████████████████████████████████████████████████████████████| 119/119 [00:00<00:00, 172.30it/s]
100%|████████████████████████████████████████████████████████████████████████████████████████████████████| 119/119 [00:00<00:00, 486.86it/s]
train loss: 0.5988994240760803
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████| 38/38 [00:00<00:00, 480.56it/s]
validation loss: 2.665893793106079
100%|████████████████████████████████████████████████████████████████████████████████████████████████████| 413/413 [00:00<00:00, 604.36it/s]
100%|████████████████████████████████████████████████████████████████████████████████████████████████████| 128/128 [00:00<00:00, 604.21it/s]
train scores: {'RMSE': 0.5988994168191223}
test scores: {'RMSE': 2.665893801278457}
```

### LINKED ISSUE #33: Dropout layers are not deterministic

### PR DESCRIPTION
**Pull request by erikhuck**
closes #26; closes #33

### CODE REVIEW

**Comment by erikhuck** [comment]
I've been able to make the DAG RNN deterministic with and without dropout. However, there is one particular case in which the DAG RNN is not deterministic. If dropout is enabled, and the LSTM layer is not bidirectional, it's deterministic. If dropout is disabled and the LSTM is bidirectional, it's deterministic. It's only when the LSTM both has dropout enabled AND it is bidirectional that there are slight deviations in RMSE. From what I've observed so far, the first two decimal places of RMSE are consistent and the third decimal place is consistently close but not always exactly the same. I've debugged and seen the outputs of the model, that's y_hat_batch, and for the most part the outputs are exactly the same each run. But occasionally you see a difference in the fourth decimal place of the outputs that is either an additional or subtraction of just 0.0001. This slight occasional difference must account for the variation of the RMSE in its third decimal place.

**Comment by erikhuck** [comment]
We previously considered making an attention system for the DAG RNN. I don't know much about attention systems, but would it possibly remove our dependence on an LSTM layer such that determinism isn't an issue?

**Comment by erikhuck** [comment]
Nevermind the above comments. The DAG RNN is now deterministic after disabling cuDNN before running it. This was accomplished by setting torch.backends.cudnn.enabled = False. The result is a slight decrease in speed but with the benefit of determinism.

**Comment by bjschoenfeld** [comment]
Thank you for trying to use the "closes #<issue nuber>" syntax. For it to work, it needs to go in the PR description, not the title.

**Comment by bjschoenfeld** [comment]
> I don't know much about attention systems, but would it possibly remove our dependence on an LSTM layer such that determinism isn't an issue?

We will not remove the LSTM models in favor of the attention models. Rather we should have both so we can compare them empirically.

File: dna/models.py [hunk 1]
```diff
@@ -633,6 +640,7 @@ def _get_primitive_name_to_enc(self, train_data):
 
         # Create a mapping of primitive names to one hot encodings
         primitive_name_to_enc = {}
+        primitive_names = sorted(primitive_names)
```
Comment (bjschoenfeld) [inline dna/models.py]: good catch!

**Comment by bjschoenfeld** [comment]
> The DAG RNN is now deterministic after disabling cuDNN before running it. This was accomplished by setting torch.backends.cudnn.enabled = False. The result is a slight decrease in speed but with the benefit of determinism.

Do you have a reference or link to help me understand why this is necessary?

**Comment by erikhuck** [comment]
> Thank you for trying to use the "closes #" syntax. For it to work, it needs to go in the PR description, not the title.

I couldn't find a way to change the PR description even after googling

**Comment by erikhuck** [comment]
> > The DAG RNN is now deterministic after disabling cuDNN before running it. This was accomplished by setting torch.backends.cudnn.enabled = False. The result is a slight decrease in speed but with the benefit of determinism.
> 
> Do you have a reference or link to help me understand why this is necessary?

https://github.com/pytorch/pytorch/issues/18110

**Comment by bjschoenfeld** [comment]
@erikhuck Can you add a test to verify the determinism? The test should look something like: run the code twice and see if they get the same results.

**Comment by erikhuck** [comment]
@bjschoenfeld Because of the several changes needed to make the tests work, I'm creating a separate pull request to make this one simpler. This one can be merged in.

**Comment by bjschoenfeld** [comment]
Ok, I created a new issue for it #66
